docs/examples/bayesian_optimisation.py

```diff
@@ -0,0 +1,737 @@
+
+# %%
+x1 = jnp.linspace(-2, 2, 100)
+x2 = jnp.linspace(-1, 1, 100)
+x1, x2 = jnp.meshgrid(x1, x2)
+x = jnp.stack([x1.flatten(), x2.flatten()], axis=1)
+y = six_hump_camel(x)
+
+fig, ax = plt.subplots(subplot_kw={"projection": "3d"})
+surf = ax.plot_surface(
+    x1,
+    x2,
+    y.reshape(x1.shape[0], x2.shape[0]),
+    linewidth=0,
+    cmap=cm.coolwarm,
+    antialiased=False,
+)
+ax.set_xlabel("x1")
+ax.set_ylabel("x2")
+plt.show()
+
+# %% [markdown]
+# For more clarity, we can generate a contour plot of the function which enables us to see
+# the global minima of the function more clearly.
+
+# %%
+x_star_one = jnp.array([[0.0898, -0.7126]])
+x_star_two = jnp.array([[-0.0898, 0.7126]])
+fig, ax = plt.subplots()
+contour_plot = ax.contourf(
+    x1, x2, y.reshape(x1.shape[0], x2.shape[0]), cmap=cm.coolwarm, levels=40
+)
+ax.scatter(
+    x_star_one[0][0], x_star_one[0][1], marker="*", color=cols[2], label="Global Minima"
+)
+ax.scatter(x_star_two[0][0], x_star_two[0][1], marker="*", color=cols[2])
+ax.set_xlabel("x1")
+ax.set_ylabel("x2")
+fig.colorbar(contour_plot)
+ax.legend()
+plt.show()
+
+# %% [markdown]
+# Next, we'll run the BO loop using Thompson sampling as before. This time we'll run the
+# experiment 5 times in order to see how the algorithm performs on average, with different
+# starting points for the initial GP model. This is good practice, as the performance
+# obtained is likely to vary between runs depending on the initialisation samples used to
+# fit the initial GP model.
+
+# %%
+lower_bound = jnp.array([-2.0, -1.0])
+upper_bound = jnp.array([2.0, 1.0])
+initial_sample_num = 5
+bo_iters = 15
+num_experiments = 5
+bo_experiment_results = []
+
+for experiment in range(num_experiments):
+    print(f"Starting Experiment: {experiment + 1}")
+    # Set up initial dataset
+    initial_x = tfp.mcmc.sample_halton_sequence(
+        dim=2, num_results=initial_sample_num, seed=key, dtype=jnp.float64
+    )
+    initial_x = jnp.array(lower_bound + (upper_bound - lower_bound) * initial_x)
+    initial_y = six_hump_camel(initial_x)
+    D = gpx.Dataset(X=initial_x, y=initial_y)
+
+    for i in range(bo_iters):
+        key, subkey = jr.split(key)
+
+        # Generate optimised posterior
+        mean = gpx.mean_functions.Zero()
+        kernel = gpx.kernels.Matern52(
+            active_dims=[0, 1], lengthscale=jnp.array([1.0, 1.0]), variance=2.0
+        )
+        prior = gpx.Prior(mean_function=mean, kernel=kernel)
+        opt_posterior = generate_optimised_posterior(D, prior, subkey)
+
+        # Draw a sample from the posterior, and find the minimiser of it
+        approx_sample = opt_posterior.sample_approx(
+            num_samples=1, train_data=D, key=subkey, num_features=500
+        )
+        x_star = optimise_sample(
+            approx_sample,
+            subkey,
+            lower_bound,
+            upper_bound,
+            num_initial_sample_points=1000,
+        )
+
```

Comment:
could even have a Thompson_Sample function that takes in a model and just return a chosen point?